File: server/r1cord_server/pipeline/instructions.py
"""INSTRUCTIONS.md text for the writer CLI."""

from __future__ import annotations

from datetime import datetime, timezone

_STYLE_HINTS = {
    "notes": (
        "Write concise working notes. Short bullets under Key points. "
        "Action items are checkable tasks, not prose."
    ),
    "minutes": (
        "Write meeting minutes: who said what that mattered, decisions, and owners. "
        "Use timestamps only when they mark a decision."
    ),
    "article": (
        "Write a short article in prose under the abstract and Key points. "
        "Action items stay a list at the end."
    ),
}


def _format_created(metadata: dict) -> str:
    raw = metadata.get("createdAt")
    if raw is None:
        return "unknown"
    try:
        ms = int(raw)
        dt = datetime.fromtimestamp(ms / 1000, tz=timezone.utc)
        return dt.strftime("%Y-%m-%d %H:%M UTC")
    except (TypeError, ValueError, OSError, OverflowError) as exc:
        raise ValueError(f"metadata.createdAt is not epoch milliseconds: {raw!r}") from exc


def _format_duration(metadata: dict) -> str:
    raw = metadata.get("durationMs")
    if raw is None:
        return "unknown"
    try:
        ms = int(raw)
    except (TypeError, ValueError) as exc:
        raise ValueError(f"metadata.durationMs is not an integer: {raw!r}") from exc
    if ms < 0:
        raise ValueError(f"metadata.durationMs is negative: {ms}")
    seconds = ms // 1000
    minutes, sec = divmod(seconds, 60)
    hours, minutes = divmod(minutes, 60)
    if hours:
        return f"{hours}h {minutes:02d}m {sec:02d}s ({ms} ms)"
    return f"{minutes}m {sec:02d}s ({ms} ms)"


def build_instructions(title: str, style: str, photos: list[str], metadata: dict) -> str:
    """Return INSTRUCTIONS.md body (under 60 lines) for the writer cwd."""
    if style not in _STYLE_HINTS:
        raise ValueError(f"unknown summary style: {style!r} (expected notes|minutes|article)")
    photo_lines: list[str]
    if photos:
        photo_lines = [
            "Place photos inline where they are relevant, with a real caption:",
            "![caption](photos/<file>)",
            "Use only these photo files (do not invent paths):",
        ]
        photo_lines.extend(f"- {name}" for name in photos)
        photo_lines.append("If a photo does not fit a point, put it after the abstract with a caption.")
    else:
        photo_lines = ["No photos. Do not add image links."]

    lines = [
        "You are summarizing an R1CORD voice recording in this working folder.",
        "",
        "Folder contents: transcript.txt (facts), transcript.json, metadata.json, photos/, this file.",
        "Write ONLY summary.md in the current working directory. Do not write any other file.",
        "",
        "Required structure of summary.md:",
        "1. First line must be exactly: [brand-header]",
        f"2. Then a heading: # {title}",
        "3. Then a one-paragraph abstract.",
        "4. Then ## Key points",
        "5. Then ## Action items",
        *photo_lines,
        "",
        "Recording:",
        f"- title: {title}",
        f"- date: {_format_created(metadata)}",
        f"- duration: {_format_duration(metadata)}",
        f"- summary style: {style}",
        "",
        f"Style rules for {style}: {_STYLE_HINTS[style]}",
        "",
        "Rules:",
        "- Do not invent facts. Use only the transcript, metadata, and listed photos.",
        "- Do not invent photo paths. Use only files listed above.",
        "- Do not append the transcript. No transcript appendix.",
        "- No timestamps unless the style is minutes and a time marks a decision.",
    ]
    text = "\n".join(lines) + "\n"
    line_count = text.count("\n")
    if line_count > 60:
        raise RuntimeError(f"INSTRUCTIONS.md is {line_count} lines; keep it under 60")
    return text

File: desktop/src/core/pipeline/instructions.js
// Port of r1cord_server/pipeline/instructions.py: INSTRUCTIONS.md for the writer CLI, a fixed
// wrapper around an editable prompt per AI review.
//
// Default prompts live here. A prompt edited on the Settings page is saved as
// `<dir of config.toml>/prompts/<kind>.md` and replaces the default until restored.

const fs = require('node:fs');
const path = require('node:path');
const { PAGE_LABELS, REVIEW_KINDS } = require('../config');
const { ValueError } = require('../errors');
const { isFile, pyInt, pyRepr, pyStrip, readText, thousands, writeText } = require('./compat');

const MAX_PROMPT_CHARS = 8_000;

const DEFAULT_PROMPTS = Object.freeze({
  summary: `Write a summary of the recording, as headings a reader can jump to from the page's Contents panel.

After the heading:
1. ## Overview: a one-paragraph abstract: what the recording is about and what came out of it.
2. ## Key points: short bullets, one idea each, the most important first. When the recording
   covers several distinct topics, group the bullets under a ### heading per topic.
3. ## Action items: a bullet list of concrete tasks, each starting with a verb. Name an owner or a
   date only when the recording does. If there are none, write "None."

Keep it concise. Use the speaker's own names for people, products and places.
`,
  outline: `Write an outline of the recording: its topics and points in the order they were discussed, as
headings a reader can jump to from the page's Contents panel.

After the heading:
- Each topic is a ## heading: a short noun phrase. Sub-topics within a topic are ### headings;
  use #### only for a further level.
- Under the lowest heading, the points made as terse bullets: fragments, not sentences; about a
  dozen words per bullet at most. Supporting detail such as numbers, names or examples goes in
  nested bullets indented by four spaces.
- Keep the order of discussion. When the speaker returns to an earlier topic, give it a new
  heading where it came up instead of moving it.
- No abstract, no commentary, no conclusions the speaker did not state.
`,
  organized: `Rewrite the recording as a clean, well-organized document that keeps ALL of its content.

After the heading:
- Group the content under clear ## headings (### where it helps), in a logical order: related
  material belongs together even when it was spoken at different times.
- Remove filler words, false starts, repetition and verbal tics ("um", "you know", "so, so").
- Keep the speaker's own wording and voice wherever possible. Fix grammar only where it gets in
  the way of reading. Keep first person if the speaker used it.
- Do not summarize or shorten: every fact, number, name, example, reason and opinion stays.
- Use paragraphs for narrative, and lists where the speaker enumerates things.
`,
});

// A prompt edit that cannot be saved (empty or too long). A ValueError, as in Python.
class PromptError extends ValueError {
  constructor(message) {
    super(message);
    this.name = 'PromptError';
  }
}

function checkKind(kind) {
  if (!REVIEW_KINDS.includes(kind)) {
    throw new ValueError(`unknown review: ${pyRepr(kind)} (expected summary, outline or organized)`);
  }
}

function promptPath(promptsDir, kind) {
  checkKind(kind);
  return path.join(String(promptsDir), `${kind}.md`);
}

function isCustom(promptsDir, kind) {
  return promptsDir !== null && promptsDir !== undefined && isFile(promptPath(promptsDir, kind));
}

// The saved override for `kind`, else the default prompt.
function loadPrompt(promptsDir, kind) {
  checkKind(kind);
  if (promptsDir !== null && promptsDir !== undefined) {
    const file = promptPath(promptsDir, kind);
    if (isFile(file)) return readText(file);
  }
  return DEFAULT_PROMPTS[kind];
}

// Store an edited prompt. Text equal to the default removes the override instead.
function savePrompt(promptsDir, kind, text) {
  checkKind(kind);
  const body = pyStrip(String(text).replace(/\r\n/g, '\n'));
  if (!body) throw new PromptError('The prompt is empty. Write the task, or use Restore default.');
  const length = Array.from(body).length;
  if (length > MAX_PROMPT_CHARS) {
    throw new PromptError(
      `The prompt is ${thousands(length)} characters; the limit is ${thousands(MAX_PROMPT_CHARS)}. Shorten it and save again.`,
    );
  }
  if (body === pyStrip(DEFAULT_PROMPTS[kind])) {
    restorePrompt(promptsDir, kind);
    return;
  }
  const file = promptPath(promptsDir, kind);
  fs.mkdirSync(path.dirname(file), { recursive: true });
  writeText(file, `${body}\n`);
}

// Delete the override so the default prompt applies again.
function restorePrompt(promptsDir, kind) {
  fs.rmSync(promptPath(promptsDir, kind), { force: true });
}

function pad2(n) {
  return String(n).padStart(2, '0');
}

function formatCreated(metadata) {
  const raw = metadata.createdAt;
  if (raw === undefined || raw === null) return 'unknown';
  const fail = () => new ValueError(`metadata.createdAt is not epoch milliseconds: ${pyRepr(raw)}`);
  let ms;
  try {
    ms = pyInt(raw);
  } catch (_error) {
    throw fail();
  }
  const date = new Date(ms);
  // datetime covers years 1..9999 only.
  if (Number.isNaN(date.getTime()) || date.getUTCFullYear() < 1 || date.getUTCFullYear() > 9999) throw fail();
  const year = String(date.getUTCFullYear()).padStart(4, '0');
  return `${year}-${pad2(date.getUTCMonth() + 1)}-${pad2(date.getUTCDate())} ${pad2(date.getUTCHours())}:${pad2(date.getUTCMinutes())} UTC`;
}

function formatDuration(metadata) {
  const raw = metadata.durationMs;
  if (raw === undefined || raw === null) return 'unknown';
  let ms;
  try {
    ms = pyInt(raw);
  } catch (_error) {
    throw new ValueError(`metadata.durationMs is not an integer: ${pyRepr(raw)}`);
  }
  if (ms < 0) throw new ValueError(`metadata.durationMs is negative: ${ms}`);
  const seconds = Math.floor(ms / 1000);
  const sec = seconds % 60;
  const totalMinutes = Math.floor(seconds / 60);
  const minutes = totalMinutes % 60;
  const hours = Math.floor(totalMinutes / 60);
  if (hours) return `${hours}h ${pad2(minutes)}m ${pad2(sec)}s (${ms} ms)`;
  return `${minutes}m ${pad2(sec)}s (${ms} ms)`;
}

/**
 * INSTRUCTIONS.md for one review: the fixed frame around the editable `prompt`.
 *
 * The frame (output file, title heading, photos, recording facts, rules) is not editable, so a
 * prompt edit cannot break the page the server publishes.
 * @param {string} kind
 * @param {string} title
 * @param {string} prompt
 * @param {string[]} photos
 * @param {object} metadata
 * @returns {string}
 */
function buildInstructions(kind, title, prompt, photos, metadata) {
  checkKind(kind);
  const name = `${kind}.md`;
  const photoLines = photos.length
    ? [
      'Place photos inline where they are relevant, with a real caption:',
      '![caption](photos/<file>)',
      'Use only these photo files (do not invent paths):',
      ...photos.map((photo) => `- ${photo}`),
      'If a photo does not fit a point, put it near the top with a caption.',
    ]
    : ['No photos. Do not add image links.'];

  const lines = [
    `You are writing the ${PAGE_LABELS[kind]} of an R1CORD voice recording in this working folder.`,
    '',
    'Folder contents: transcript.txt (facts), transcript.json, metadata.json, photos/, this file.',
    `Write ONLY ${name} in the current working directory. Do not write any other file.`,
    '',
    `${name} must start with a heading: # ${title}`,
    'Everything after the heading follows the task below.',
    '',
    '## Task',
    '',
    pyStrip(prompt),
    '',
    '## Photos',
    '',
    ...photoLines,
    '',
    '## Recording',
    '',
    `- title: ${title}`,
    `- date: ${formatCreated(metadata)}`,
    `- duration: ${formatDuration(metadata)}`,
    '',
    '## Rules (always apply, whatever the task says)',
    '',
    '- Do not invent facts. Use only the transcript, metadata, and listed photos.',
    '- Do not invent photo paths. Use only files listed above.',
    '- Do not append the transcript. No transcript appendix.',
    `- Keep the heading of ${name} as described above.`,
  ];
  return `${lines.join('\n')}\n`;
}

module.exports = {
  DEFAULT_PROMPTS,
  MAX_PROMPT_CHARS,
  PromptError,
  buildInstructions,
  isCustom,
  loadPrompt,
  promptPath,
  restorePrompt,
  savePrompt,
};
